jetbrains-core/src-211+/software.aws.toolkits.jetbrains/services/ssm/SsmPlugin.kt

```diff
@@ -0,0 +1,114 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import com.intellij.execution.configurations.GeneralCommandLine
+import com.intellij.execution.util.ExecUtil
+import com.intellij.openapi.progress.ProgressIndicator
+import com.intellij.openapi.util.SystemInfo
+import com.intellij.util.io.Decompressor
+import com.intellij.util.system.CpuArch
+import com.intellij.util.system.CpuArch.isArm64
+import software.aws.toolkits.core.utils.getLogger
+import software.aws.toolkits.jetbrains.core.getTextFromUrl
+import software.aws.toolkits.jetbrains.core.saveFileFromUrl
+import software.aws.toolkits.jetbrains.core.tools.FourPartVersion
+import software.aws.toolkits.jetbrains.core.tools.ManagedToolType
+import software.aws.toolkits.jetbrains.core.tools.Tool
+import software.aws.toolkits.jetbrains.core.tools.ToolType
+import software.aws.toolkits.jetbrains.core.tools.VersionRange
+import software.aws.toolkits.jetbrains.core.tools.until
+import software.aws.toolkits.jetbrains.utils.checkSuccess
+import java.nio.file.Files
+import java.nio.file.Path
+import java.time.Duration
+import kotlin.streams.asSequence
+
+object SsmPlugin : ManagedToolType<FourPartVersion> {
+    private val isUbuntu by lazy {
+        ExecUtil.execAndGetOutput(GeneralCommandLine("uname", "-a"), VERSION_TIMEOUT.toMillis().toInt()).stdout.contains("Ubuntu", ignoreCase = true)
+    }
+    override val id: String = "SSM-Plugin"
+    override val displayName: String = "AWS Session Manager Plugin"
+
+    override fun supportedVersions(): VersionRange<FourPartVersion> = FourPartVersion(1, 2, 0, 0) until FourPartVersion(2, 0, 0, 0)
+
+    override fun determineVersion(path: Path): FourPartVersion {
+        val processOutput = ExecUtil.execAndGetOutput(
+            GeneralCommandLine(path.toString(), "--version"),
+            VERSION_TIMEOUT.toMillis().toInt()
+        )
+
+        if (!processOutput.checkSuccess(LOGGER)) {
+            throw IllegalStateException("Failed to determine version of $displayName")
+        }
+
+        return FourPartVersion.parse(processOutput.stdout.trim())
+    }
+
+    override fun downloadVersion(version: FourPartVersion, destinationDir: Path, indicator: ProgressIndicator?): Path {
+        // TODO: Move to CpuArch, FIX_WHEN_MIN_IS_211
+        val downloadUrl = when {
+            SystemInfo.isWindows -> windowsUrl(version)
+            SystemInfo.isMac -> macUrl(version)
+            SystemInfo.isLinux && isUbuntu && isArm64() -> ubuntuArm64Url(version)
+            SystemInfo.isLinux && isUbuntu && CpuArch.isIntel64() -> ubuntuI64Url(version)
```

Comment:
for this - lets not copy the *whole* file, lets create a seprate file, that has *just* the required compatibility functions in it.

Reply:
This was merged with : https://github.com/aws/aws-toolkit-jetbrains/pull/2912